### PR TITLE
rust: Migrate to latest revision of `zcash/librustzcash` crates

### LINKED ISSUE #6566: Use extracted sapling-crypto crate in zcashd

### LINKED ISSUE #6806: Use latest librustzcash crates (zcash_primitives 0.14, etc.) in zcashd

### PR DESCRIPTION
**Pull request by str4d**
fixes #6566, fixes #6806

### CODE REVIEW

File: src/rust/src/tests/signatures.rs [hunk 1]
```diff
@@ -481,30 +480,20 @@ fn redjubjub_signatures() {
...
 
-        assert!(vk.verify(&tv.m, &sig, SPENDING_KEY_GENERATOR));
-        assert!(rvk.verify(&tv.m, &rsig, SPENDING_KEY_GENERATOR));
-        assert!(!vk.verify(&tv.m, &rsig, SPENDING_KEY_GENERATOR));
-        assert!(!rvk.verify(&tv.m, &sig, SPENDING_KEY_GENERATOR));
+        assert_eq!(vk.verify(&tv.m, &sig), Ok(()));
+        assert_eq!(rvk.verify(&tv.m, &rsig), Ok(()));
+        assert!(matches!(vk.verify(&tv.m, &rsig), Err(_)));
+        assert!(matches!(rvk.verify(&tv.m, &sig), Err(_)));
     }
```
Comment (str4d) [inline src/rust/src/tests/signatures.rs:498]: Something about this change from `zcash_primitives::sapling::redjubjub` (which is being removed in https://github.com/zcash/librustzcash/pull/1056) to the `redjubjub` crate is causing the test vectors to fail.

Comment (str4d) [reply src/rust/src/tests/signatures.rs:498]: Aha: the problem is that the test vectors were written before key prefixing was added to RedJubjub. `zcash_primitives::sapling::redjubjub` was never updated to include key prefixing, instead leaving it to the callsites. The `redjubjub` crate correctly includes key prefixing, but that means the old test vectors cannot be used with it. I'll open a PR to regenerate those test vectors.

**Review by ebfull — COMMENTED** [review]
As part of this PR we probably should also modify `contrib/debian/copyright` to change orchard's license away from BOSL. (Are there other remnants of BOSL in our codebase after this?)

**Review by ebfull — APPROVED** [review]

File: src/rust/src/tests/signatures.rs [hunk 2]
```diff
@@ -50,18 +47,18 @@ fn redjubjub_signatures() {
...
-                0xea, 0xa0, 0x57, 0x47, 0x6b, 0x4a, 0xb4, 0x82, 0x28, 0x8b, 0x93, 0xdf, 0x8f, 0xe0,
-                0xc5, 0xce, 0x9d, 0x78, 0x83, 0x67, 0xf2, 0xbe, 0x55, 0x1b, 0x7f, 0x7a, 0x82, 0xa6,
-                0xdb, 0x36, 0x04, 0x68, 0xde, 0xb9, 0xa7, 0xb7, 0xaf, 0xaa, 0xdf, 0xec, 0xa6, 0xf4,
-                0x81, 0x19, 0x3d, 0xc6, 0x57, 0x57, 0x47, 0xf6, 0x0a, 0x1a, 0x8a, 0x48, 0xff, 0x0a,
-                0xd7, 0x0c, 0xf8, 0xcb, 0x8d, 0x52, 0x8e, 0x08,
+                0xdc, 0xa3, 0xbb, 0x2c, 0xb8, 0xf0, 0x48, 0xcc, 0xab, 0x10, 0xae, 0xd7, 0x75, 0x46,
+                0xc1, 0xdb, 0xb1, 0x0c, 0xc4, 0xfb, 0x15, 0xab, 0x02, 0xac, 0xae, 0xf9, 0x44, 0xdd,
+                0xab, 0x8b, 0x67, 0x22, 0x54, 0x5f, 0xda, 0x4c, 0x62, 0x04, 0x6d, 0x69, 0xd9, 0x8f,
+                0x92, 0x2f, 0x4e, 0x8c, 0x21, 0x0b, 0xc4, 0x7b, 0x4f, 0xdd, 0xe0, 0xa1, 0x94, 0x71,
+                0x79, 0x80, 0x4c, 0x1a, 0xce, 0x56, 0x90, 0x05,
```
Comment (daira) [inline src/rust/src/tests/signatures.rs:54]: Why have these test vectors changed?

File: src/rust/src/wallet.rs [hunk 3]
```diff
@@ -150,7 +149,8 @@ pub struct Wallet {
     nullifiers: BTreeMap<Nullifier, OutPoint>,
     /// The incremental Merkle tree used to track note commitments and witnesses for notes
     /// belonging to the wallet.
-    commitment_tree: BridgeTree<MerkleHashOrchard, u32, NOTE_COMMITMENT_TREE_DEPTH>,
+    // TODO: Replace this with an `orchard` crate constant (they happen to be the same).
+    commitment_tree: BridgeTree<MerkleHashOrchard, u32, { sapling::NOTE_COMMITMENT_TREE_DEPTH }>,
```
Comment (daira) [inline src/rust/src/wallet.rs:153]:
```suggestion
    commitment_tree: BridgeTree<MerkleHashOrchard, u32, 32>,
```
Temporarily, it is better to hard-code than to use another constant that happens to be the same (also change the comment and specify the type of the constant if necessary).

File: src/rust/src/wallet.rs [hunk 4]
```diff
@@ -1370,7 +1370,10 @@ pub extern "C" fn orchard_wallet_load_note_commitment_tree(
 #[no_mangle]
 pub extern "C" fn orchard_wallet_init_from_frontier(
     wallet: *mut Wallet,
-    frontier: *const bridgetree::Frontier<MerkleHashOrchard, NOTE_COMMITMENT_TREE_DEPTH>,
+    frontier: *const bridgetree::Frontier<
+        MerkleHashOrchard,
+        { sapling::NOTE_COMMITMENT_TREE_DEPTH },
```
Comment (daira) [inline src/rust/src/wallet.rs:1375]: Same here.

File: src/test/coins_tests.cpp [hunk 5]
```diff
@@ -306,9 +306,10 @@ class TxWithNullifiers
 
         // The Orchard bundle builder always pads to two Actions, so we can just
         // use an empty builder to create a dummy Orchard bundle.
+        // TODO: With the new BundleType::DEFAULT this is no longer true. Fix this.
```
Comment (daira) [inline src/test/coins_tests.cpp:309]: Please file an issue.

File: src/wallet/gtest/test_wallet.cpp [hunk 6]
```diff
@@ -2046,21 +2049,21 @@ TEST(WalletTests, UpdatedSaplingNoteData) {
     auto m = GetTestMasterSaplingSpendingKey();
 
     // Generate dummy Sapling address
-    auto sk = m.Derive(0);
+    auto sk = m.Derive(0 | HARDENED_KEY_LIMIT);
```
Comment (daira) [inline src/wallet/gtest/test_wallet.cpp:2052]: This constant only happens to have the right value. (Same comment on other instances.)

**Review by daira — APPROVED** [review]
utACK with comments.

**Comment by ebfull** [comment]
Force pushed so we can get CI running on this.

**Review by nuttycom — APPROVED** [review]
utACK the changes [since the approving reviews](https://github.com/zcash/zcash/compare/b0ac5c38454b0767feb1a8cc67279332fd0992b1..fd675c3205fbd121e659297f216ebc94d95cb074)

**Pull request merged**